### ISSUE TITLE
Socket connection hangs on Linux when buffers aren't read fully before closing connection

### ISSUE THREAD

**Issue by aviviadi** [dotnet/corefx#17330]
On windows 10 works perfectly. On ubuntu16 hangs every time. dotnet 1.0.0-preview2-1-003177

command to reproduce (every time) at the following links :
```
dotnet test src/FastTests -class FastTests.Server.OAuth.CanAuthenticate -verbose
```

At first I used WebSocket and the connection hanged. Repro every time at:
https://github.com/ravendb/ravendb/commit/1990e41e4da8d3b97e4ba98714028ac27eb9f514

While thinking this is a websocket problem, I changed the code to use http requests, but it still hangs in the same way.  Only when I drains the buffers it passes:
https://github.com/ravendb/ravendb/commit/d8f4aec4561e51db50b739288b520b8121e45a22

The last commit needed only in Linux. In windows it works fine.

My suspect is websocket disconnections timeouts and hangs on linux are related to the above bug, and solving it for http requests will help also with websockets.

**Comment by stephentoub** [comment]
Can you share an isolated repro that doesn't involve ravendb?  On Unix, the Websockets implementation in corefx is entirely separate from the HttpClient/HttpWebRequest, with the former implemented on Sockets and the latter on libcurl, so it's possible but unlikely there's a direct connection.

**Comment by aviviadi** [comment]
I understand the difference you mentioned, so it might be a different problem in WebSockets, although it hangs in the same point in the same scenario.

I've tried to repro outside the project with no success (also with xunit no success). So no isolated repro is available.. 

I did try to debug it using Visual Studio 2017 (remote ssh), and also with tcpdump. It showed clearly it sends a response which never reach the target if the read buffer aren't drained.

**Comment by stephentoub** [comment]
> I've tried to repro outside the project with no success

Thanks for trying.  How do you know it's a problem with the .NET libs rather than with the ravendb code above it?

**Comment by aviviadi** [comment]
This is simple client-server request and response.  The hang occurs when the above server's BL wants to reply to the client with a response (error string of somekind) before reading the client request body.  That's all..    If I drain the read buffers it works.    And it works with and without draining the buffer in Windows.

So I think this is narrowing the problem to Linux & Sockets (in parallel xunit tests)

**Comment by stephentoub** [comment]
> This is a simple client-server request and response
> That's all

But then it should be possible to come up with an isolated repro, right?  There are potentially many differences that could be causing an issue and that would explain the difference between it working on Windows and not working on Linux, including for example Unix-specific code in ravendb, or fundamental differences at the OS level of how sockets work.  And the fact that it repros on two completely different networking stack implementations on Linux (one that uses libcurl and one that uses .NET Sockets) suggests that the problem is either a layer above the .NET libs or a layer below, though like you say the fact that it hangs in exactly the same place could just be coincidence.  It'd be very helpful to get an isolated repro; without that, making progress will be difficult, as the request would seem to be to debug ravendb?

**Comment by stephentoub** [comment]
Ah, wait, I looked at your commit.  It's dealing with `HttpContext`.  Is that the ASP.NET HttpContext?

**Comment by aviviadi** [comment]
No, it is kestrel's.   Ok.. Now we are able to reproduce it without our project : https://gist.github.com/anonymous/af69718f23acf2e4eaff33ff36bf109e

The scenario is that we write from the response without reading the complete data from the request.
Setup a Kesterl app with this code and then run the client code, typically within 3 - 5 iterations, this will cause a hang.

**Comment by aviviadi** [comment]
Important to note that while this is a repro with plain requests, but we saw the exact same behavior when using web sockets as well, and I strongly suspect that this is related.

**Comment by stephentoub** [comment]
> No, it is kestrel's

Yeah, that's what I meant by ASP.NET; sorry for confusion.

When I said that Websockets and HTTP implementations were unrelated, that was when I thought you were talking about client-side / making outbound requests.  If you're talking about server-side / accepting requests using kestrel, that's an entirely different story.

So, just to be clear, the hang is occurring for you in the server-side of this, correct?  Meaning, for example, in the repro you shared, you could be using any client and it would cause a hang in the server?

Assuming yes, sounds like an issue with kestrel (which uses libuv for networking), and we should move the issue to https://github.com/aspnet/kestrelhttpserver.

**Comment by aviviadi** [comment]
We believe that the issue is on the client side, since we can kill the Kestrel server process and on the client side, it doesn't abort, but have to wait for the full timeout, in this case, the OS already closed all sockets, so I wouldn't expect it.

In addition, tracing via tcpdump shows that the response has been sent to the client.

**Comment by stephentoub** [comment]
I see.  So to summarize:

- The client posts a request to the server.
- The server doesn't read any the request body and instead just sends a response.
- As a result, the client hangs.
- And because the client doesn't disconnect, the server doesn't disconnect / keeps the connection open and assigned to that client.

Have I understood correctly now?

What OS is the client on?  What version of libcurl is installed?

**Comment by stephentoub** [comment]
@aviviadi, does this repro for you if you change the status code in the response from 333 to 200?

I think I may know what's happening.  The client starts sending the request.  As soon as the request starts to be handled by kestrel, you send back an error code 333.  When libcurl gets back that error code, it stops sending any more of the request body.  If you look at a trace output from the client, you can see this:
```
[Microsoft-System-Net-Http-6] (2, 101, CurlDebugFunction, CURLINFO_HEADER_IN: HTTP/1.1 333 Unknown).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveHeadersCallback, Size: 22).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveHeadersCallback, Received status line).
[Microsoft-System-Net-Http-6] (2, 101, CurlDebugFunction, CURLINFO_HEADER_IN: Date: Tue, 21 Mar 2017 11:37:43 GMT).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveHeadersCallback, Size: 37).
[Microsoft-System-Net-Http-6] (2, 101, CurlDebugFunction, CURLINFO_HEADER_IN: Server: Kestrel).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveHeadersCallback, Size: 17).
[Microsoft-System-Net-Http-6] (2, 101, CurlDebugFunction, CURLINFO_HEADER_IN: Transfer-Encoding: chunked).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveHeadersCallback, Size: 28).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveHeadersCallback, Size: 2).
[Microsoft-System-Net-Http-6] (2, 101, CurlDebugFunction, CURLINFO_TEXT: HTTP error before end of send, stop sending).
[Microsoft-System-Net-Http-6] (2, 101, CurlDebugFunction, CURLINFO_DATA_IN: 1031 bytes).
[Microsoft-System-Net-Http-6] (2, 101, CurlReceiveBodyCallback, Size: 1024).
[Microsoft-System-Net-Http-6] (2, 101, EnsureResponseMessagePublished, Publishing response message).
[Microsoft-System-Net-Http-6] (2, 101, ReadAsync, Buffer: 4096, Offset: 0, Count: 4096).
[Microsoft-System-Net-Http-6] (2, 101, RequestUnpause, ).
[Microsoft-System-Net-Http-6] (2, 0, RequestWakeup, ).
[Microsoft-Diagnostics-DiagnosticSource-2] (HttpHandlerDiagnosticListener, System.Net.Http.Response, System.Object[]).
[Microsoft-System-Net-Http-6] (2, 101, TransferDataToResponseStream, Length: 1024).
[Microsoft-System-Net-Http-6] (2, 101, TransferDataToResponseStream, Completing pending read with 1024 bytes).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait wake-up).
[Microsoft-System-Net-Http-6] (2, 101, HandleIncomingRequests, Type: Unpause).
[Microsoft-System-Net-Http-6] (2, 101, ReadAsync, Buffer: 4096, Offset: 0, Count: 4096).
[Microsoft-System-Net-Http-6] (2, 101, RequestUnpause, ).
[Microsoft-System-Net-Http-6] (2, 0, RequestWakeup, ).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait wake-up).
[Microsoft-System-Net-Http-6] (2, 101, HandleIncomingRequests, Type: Unpause).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait timeout).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait timeout).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait timeout).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait timeout).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait timeout).
[Microsoft-System-Net-Http-6] (2, 0, WaitForWork, Wait timeout).
```
In particular that "HTTP error before end of send, stop sending)." appears to be key.  And if you look at a corresponding log on the server side, you see output like this:
```
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 0.1799ms 333
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 POST http://localhost:5000/ text/plain; charset=utf-8 1152
info: Microsoft.AspNetCore.Server.Kestrel[17]
      Connection id "0HL3G9QO0VQRK" bad request data: "Unexpected end of request content."
Microsoft.AspNetCore.Server.Kestrel.BadHttpRequestException: Unexpected end of request content.
```
so it appears that when the race condition causes the client to stop sending the body, it and the server then get into a deadlock waiting for the other to finish.

If I change the response code to 200 instead of 333, the hang goes away.

This would also explain why you don't see the issue on Windows: the WinHTTP implementation sends the full body of the request before processing any of the response.

**Comment by aviviadi** [comment]
Yes you've understood correctly :)  (I never requested you to debug ravendb...)

OS : ubuntu 16.04 (kernel 4.4), curl 7.47

**Comment by aviviadi** [comment]
I tried with returning 201 and it works.  500 or other non IsSuccessfull rc doesn't

**Comment by stephentoub** [comment]
@aviviadi, from what I can see from the trace as well, it appears chunked transfer-encoding is being used, and the client isn't receiving the necessary 0-length chunk at the end.  Do you see that in your tcpdump trace?  If it's not being received, that would explain why the client is sitting there waiting.  And maybe that's not being sent because the server incurs a BadHttpRequestException exception due to it issuing an error code that causes the client to stop sending the request?

**Comment by aviviadi** [comment]
This is how the last successful response looks like:
```
13:56:11.570338 IP 127.0.0.1.49880 > 127.0.0.1.5000: tcp 170
E.....@.@..J...............<.N.............
.
...
..POST / HTTP/1.1
Host: localhost:5000
Accept: */*
Accept-Encoding: gzip,deflate
Content-Type: text/plain; charset=utf-8
Content-Length: 1152
Expect: 100-continue


13:56:11.570502 IP 127.0.0.1.5000 > 127.0.0.1.49880: tcp 1151
E.....@.@.`..............N.........    .......
.
...
..HTTP/1.1 500 Internal Server Error
Date: Tue, 21 Mar 2017 11:56:11 GMT
Transfer-Encoding: chunked
Server: Kestrel

400
................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
```

After that the server tries to send as before :
```
13:56:11.608594 IP 127.0.0.1.49880 > 127.0.0.1.5000: tcp 0
E..4..@.@....................N.......(.....
.
.$.
..
```
and then it gets into the above mentioned deadlock

**Comment by stephentoub** [comment]
I can't tell from that output whether the ending "0\r\n\r\n" chunk is being sent in the case that causes the hang.  Can you?

**Comment by aviviadi** [comment]
ohh you mean \0.  I thought your where looking for headers["Content-Length"] = 0

**Comment by stephentoub** [comment]
I meant, the response is using chunked transfer encoding, so the client only knows that it's received everything when it receives the terminating chunk, and at least from what I've been able to repro locally, it looks like that terminating chunk isn't being sent.  Are you seeing the same thing?

**Comment by aviviadi** [comment]
First here is the trace:
```
	0x00b0:  0a00 0000 0000 0000 0000 0000 0000 0000  ................
	0x00c0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x00d0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x00e0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x00f0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0100:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0110:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0120:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0130:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0140:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0150:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0160:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0170:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0180:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0190:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x01a0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x01b0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x01c0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x01d0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x01e0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x01f0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0200:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0210:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0220:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0230:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0240:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0250:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0260:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0270:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0280:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0290:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x02a0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x02b0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x02c0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x02d0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x02e0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x02f0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0300:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0310:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0320:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0330:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0340:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0350:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0360:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0370:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0380:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0390:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x03a0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x03b0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x03c0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x03d0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x03e0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x03f0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0400:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0410:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0420:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0430:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0440:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0450:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0460:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0470:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0480:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x0490:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x04a0:  0000 0000 0000 0000 0000 0000 0000 0000  ................
	0x04b0:  000d 0a                                  ...
13:56:11.608594 IP 127.0.0.1.49880 > 127.0.0.1.5000: tcp 0
	0x0000:  4500 0034 b4ce 4000 4006 87f3 7f00 0001  E..4..@.@.......
	0x0010:  7f00 0001 c2d8 1388 079e cae6 f64e f8ff  .............N..
	0x0020:  8010 01f8 fe28 0000 0101 080a 1b0a 0724  .....(.........$
	0x0030:  1b0a 071a                                ....
```

**Comment by aviviadi** [comment]
Should I look for a complete new response from server to client with Content-Length == 0 ? 
I am not sure I follow

**Comment by aviviadi** [comment]
if you mean the ending "0\r\n\r\n"  then it isn't there.  As we see there's 0\r\n only in the end :
```
0x04b0:  000d 0a        
```

**Comment by stephentoub** [comment]
@cesarbs, are there known issues in kestrel here? As a summary, here's what I'm seeing:
1. The client starts sending the request.
2. The server implementation doesn't read the request, and as in @aviviadi's example, just starts sending a response, including an error code.
3. Since the server didn't wait to read the full request body, the client is still sending it.  The client receives the response error status, and shuts down the sending connection.
4. kestrel sees that no more data will be arriving in the request, which doesn't match the content-length it was expecting, and throws a BadHttpRequestException "Unexpected end of request content".  It then looks like it doesn't send a terminating chunk, nor does it close the connection.
5. The client waits for the response to be completed, but it waits "forever", as no terminating chunk was sent nor was the connection closed.

(Note that libcurl added in the last 6 months, in 7.51, an option https://curl.haxx.se/libcurl/c/CURLOPT_KEEP_SENDING_ON_ERROR.html so that it won't stop sending request data if it receives an error status from the server, so we could consider turning that on, but a) that would only affect very recent usage, and b) there are likely many clients out there on the net today with this behavior, so if I'm right about the chunk terminator, it seems like there's a fix to be made in kestrel as well.)

**Comment by stephentoub** [comment]
@aviviadi, yes.  To make it easier to see, I wrote a simple Socket-based client that just shuts down the sending connection once it receives the status line:
```C#
using System;
using System.IO;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using System.Text;

class Program
{
    static void Main(string[] args) => MainAsync().GetAwaiter().GetResult();
    static async Task MainAsync()
    {
        using (Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
        {
            await s.ConnectAsync(IPAddress.Loopback, 5000);
            await s.SendAsync(new ArraySegment<byte>(Encoding.ASCII.GetBytes("POST / HTTP/1.1\r\nContent-Length: 123\r\n\r\nnot123bytes")), SocketFlags.None);

            using (NetworkStream ns = new NetworkStream(s))
            {
                int c;
                while ((c = ns.ReadByte()) != '\r');
                s.Shutdown(SocketShutdown.Send);

                int col = 0;
                while ((c = ns.ReadByte()) >= 0)
                {
                    Console.Write("0x" + c.ToString("x2") + " ");
                    if ((++col % 8) == 0) Console.WriteLine();
                }
            }
       }
       Console.WriteLine();
    }
}
```
I also changed your repro to output data other than lots of 0s to make it easier to see what's going on:
```C#

        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
            loggerFactory.AddConsole();
            app.Run(ctx => {
                byte[] buffer = Enumerable.Range(0, 256).SelectMany(i => new byte[] { 0xDE, 0xAD, 0xBE, 0xEF }).ToArray();
                ctx.Response.StatusCode = 200;
                ctx.Response.Body.Write(buffer, 0, 1024);
                ctx.Response.Body.Flush();
                return Task.CompletedTask;
            });
        }
```
When I run this, the end of the client output looks like:
```
0xbe 0xef 0xde 0xad 0xbe 0xef 0xde 0xad
0xbe 0xef 0xde 0xad 0xbe 0xef 0xde 0xad
0xbe 0xef 0xde 0xad 0xbe 0xef 0xde 0xad
0xbe 0xef 0x0d 0x0a
```
but if I change the client repro to output the right number of bytes:
```C#
Encoding.ASCII.GetBytes("POST / HTTP/1.1\r\nContent-Length: 7\r\n\r\n7 bytes")
```
it outputs at the end:
```
0xbe 0xef 0xde 0xad 0xbe 0xef 0xde 0xad
0xbe 0xef 0x0d 0x0a 0x30 0x0d 0x0a 0x0d
0x0a
```
That 0x30 0x0d 0x0a 0x0d 0x0a is the text:
```
0\r\n
\r\n
```
i.e. the terminating 0-length chunk.

In other words, because the server tells the client there's an error, the client shuts down the send side of the connection, the server sees that as an unexpected end of content, and apparently doesn't send the terminating chunk, which then leaves the client waiting indefinitely for the end of the response.

**Comment by cesarbs** [comment]
@stephentoub 

> kestrel sees that no more data will be arriving in the request, which doesn't match the content-length it was expecting, and throws a BadHttpRequestException "Unexpected end of request content". It then looks like it doesn't send a terminating chunk, nor does it close the connection.

Looks like a bug in Kestrel. The exception is expected since Kestrel will consider a FIN mid-request as a connection abort. But it should have closed the connection to the client.

Will look deeper once I set up the repro on my side.

**Comment by stephentoub** [comment]
Thanks for investigating, @cesarbs.  It's possible I've analyzed something incorrectly as well, but hopefully I've not sent you on a wild goose chase :)

**Comment by karelz** [comment]
Are we ready to move the bug to Kestrel repo? Or is there anything left for CoreFX to track? (disclaimer: I just skimmed few last posts)

**Comment by stephentoub** [comment]
> Are we ready to move the bug to Kestrel repo?

No, not until @cesarbs completes his investigation.

> Or is there anything left for CoreFX to track?

There's the potential for HttpClient on libcurl to set its https://curl.haxx.se/libcurl/c/CURLOPT_KEEP_SENDING_ON_ERROR.html if on a new enough version.  We need to decide if we want to do that.  I'm not sure it's a good idea, but we could discuss it.

**Comment by cesarbs** [comment]
https://github.com/aspnet/KestrelHttpServer/issues/1530#issuecomment-288177478

**Comment by cesarbs** [comment]
@aviviadi: @stephentoub and I have finished investigating this issue, and here's what's happening:

1. The client (libcurl) starts sending the request.
2. Kestrel sends an error response.
3. The client sees the error response and stops sending the request (this is a libcurl behavior), but does NOT close the sending side of the connection (HTTP clients typically wait until they receive the entire response before closing the connection, because most HTTP servers treat a mid-request FIN like a connection reset).
4. After the application runs, [Kestrel tries to consume the rest of the request body from the client](https://github.com/aspnet/KestrelHttpServer/blob/rel/1.1.1/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs#L148). But the client has stopped sending the request, so Kestrel hangs here.
5. The client hangs waiting for the chunked terminator from the server, [which is only sent after it consumes the rest of the request body](https://github.com/aspnet/KestrelHttpServer/blob/rel/1.1.1/src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs#L154) (we'll work on improving this behavior).

There are two ways to mitigate this on the app's side: either don't chunk the error response (so libcurl won't wait for the chunk terminator), or send a `Connection: close` header in it. A `Connection: close` header lets Kestrel know the connection is about to close, so it won't attempt to read the rest of the request body. It will not hang and will send the chunked terminator, and libcurl won't hang waiting for it.

**Comment by ayende** [comment]
@cesarbs Will sending `Connection: Close` header cause the TCP connection to be dropped?

**Comment by davidfowl** [comment]
@ayende yes, it will

**Comment by stephentoub** [comment]
@cesarbs, thanks for the help investigating, and for the fix in https://github.com/aspnet/KestrelHttpServer/issues/1537.  We also discussed whether we should use the new libcurl CURLOPT_KEEP_SENDING_ON_ERROR flag if it's available, and decided that doing so is not a good idea (e.g. continuing to send large post data even if the server has already said it's not interested, such as due to a redirect or auth or some other error status).  As such, I'm going to close this out.
